mesonbuild/interpreter.py

```diff
@@ -3657,8 +3657,16 @@ def func_configure_file(self, node, args, kwargs):
         # Install file if requested, we check for the empty string
         # for backwards compatibility. That was the behaviour before
         # 0.45.0 so preserve it.
-        idir = kwargs.get('install_dir', None)
-        if isinstance(idir, str) and idir:
+        install = kwargs.get('install', None)
```

Comment:
why don't we default to `false` here? It looks like you're trying to account for install not being set, but install_dir being set?

If that's the case I'd rather see something like
```python
idir = ...
install = kwargs.get('install')
if install is None:
  if idir:
    install = True
    mlog.deprecation('DEPRECATION: install_dir set but install is not set, assuming install is true.')
  else:
    install = False
```
That way we don't have to deal with install being `None` in the rest of the function.

Reply:
Not sure we should warn here.

I would go with something like this:
```
        idir = kwargs.get('install_dir', '')
        if not isinstance(idir, str):
            raise InterpreterException('"install_dir" must be a string')
        install = kwargs.get('install', idir != '')
        if not isinstance(install, bool):
            raise InterpreterException('"install" must be a boolean')
        if install:
            if not idir:
                raise InterpreterException('"install_dir" must be specified when "install" is true')
             ....
```

Reply:
I prefer the error myself, just not sure how important it is to keep backwards compatibility in this case.

Reply:
Although in your example `install_dir` implicitly sets `install`, do we do that in other cases of install+install_dir?

Reply:
I'm previous versions of meson, install kwarg was not documented, so there is no compatibility issues whatever we do now.

In other places install_dir doesn't imply install:true but here we have to do that for compatibility.

Reply:
right, in my version setting `install_dir` but not `install` would have a warning, but yours catches `install` but not `install_dir` and errors.

Reply:
To be clear, I think we should have both, a deprecation warning about `install_dir` but not `install`, and a hard error for `install` but not `install_dir`.

Reply:
Mh, I'm not sure about such deprecation warning either, as when `install_dir` is set but not `install` is set doesn't cause any issue.

Also to me is wrong, because there might be cause where you've set `install` to be dependent on another boolean, for example in a loop, or some other cases (like here https://github.com/mesonbuild/meson/issues/4160#issue-359305781).

If by `not install` you instead mean that in the case install isn't provided at all, then it could be discussed.